Data/DataModel/PuzzleStatePerTeam.cs

```diff
@@ -68,5 +73,124 @@ public bool IsSolved
         /// Notes input by the team for this puzzle
         /// </summary>
         public string Notes { get; set; }
+
+        #region Consistency Helpers
+        public static async Task<bool> EnsureStateForPuzzleAsync(PuzzleServerContext context, int eventId, int puzzleId, bool saveChanges = true)
+        {
+            var teamsQ = context.Teams.Where(team => team.Event.ID == eventId).Select(team => team.ID);
+            var puzzleStateTeamsQ = context.PuzzleStatePerTeam.Where(state => state.PuzzleID == puzzleId).Select(state => state.TeamID);
+            var teamsWithoutState = await teamsQ.Except(puzzleStateTeamsQ).ToListAsync();
+
+            if (teamsWithoutState.Count > 0)
+            {
+                for (int i = 0; i < teamsWithoutState.Count; i++)
+                {
+                    context.PuzzleStatePerTeam.Add(new DataModel.PuzzleStatePerTeam() { PuzzleID = puzzleId, TeamID = teamsWithoutState[i] });
+                }
+
+                if (saveChanges)
+                {
+                    await context.SaveChangesAsync();
```

Comment:
I think you might have handle it correctly in the code so far, but I'm a bit nervous about correct save handling. Is there a good way to encapsulate this to avoid errors?

Reply:
Even a transaction won't work if one person adds a puzzle at (nearly) the same time that someone else adds a team. You've convinced me that I have no idea how to successfully do this, so I am removing all the perf-related parts of this change. The player page doesn't need to ensure that state exists, since it only needs to display puzzles that are unlocked and you can't unlock without having a record.

Any code that unlocks a puzzle via prerequisites will have to find its own way to ensure state. Eventually we should make a helper for this.

The status page will still need to enforce that state exists unless someone can show me how to create a join that includes missing items - I cannot figure out how to do this despite looking at several examples.

Reply:
I'd suggest that we still have code that creates these when puzzles and teams are created. As long as it only calls ```SaveChangesAsync``` once, it will ensure there are no partially built states when there isn't a race condition.

Reply:
That is exactly what the code you reviewed was doing. I am lost here.

Reply:
Specifically, Create.cs passed saveChanges:false so that there would be a single SaveChangesAsync call. But you've convinced me that this isn't solid enough because a new team could have snuck in at the same time, so I don't know that it matters.